# Qt 5.15.2 iOS UIScene backport for Xcode 27
# Based on Qt's upstream iOS scene lifecycle change (qtbase commit 76ebf51bc08f6af624a8540e7af88b9129b22ae1).
# Apply from the qtbase source root with: git apply -p1 qios-uiscene.patch

diff --git a/src/plugins/platforms/ios/qiosapplicationdelegate.mm b/src/plugins/platforms/ios/qiosapplicationdelegate.mm
--- a/src/plugins/platforms/ios/qiosapplicationdelegate.mm
+++ b/src/plugins/platforms/ios/qiosapplicationdelegate.mm
@@ -39,17 +39,101 @@
 
 #include "qiosapplicationdelegate.h"
 
+#include "qiosglobal.h"
 #include "qiosintegration.h"
 #include "qiosservices.h"
 #include "qiosviewcontroller.h"
 #include "qioswindow.h"
+#include "qiosscreen.h"
 
 #include <qpa/qplatformintegration.h>
 
 #include <QtCore/QtCore>
 
+@interface QIOSWindowSceneDelegate : NSObject<UIWindowSceneDelegate>
+@property (nonatomic, retain) QUIWindow *window;
+- (void)configureWindowForScene:(UIWindowScene *)windowScene;
+@end
+
 @implementation QIOSApplicationDelegate
 
+- (UISceneConfiguration *)application:(UIApplication *)application
+                          configurationForConnectingSceneSession:(UISceneSession *)connectingSceneSession
+                          options:(UISceneConnectionOptions *)options
+{
+    Q_UNUSED(application);
+    qCDebug(lcQpaWindowScene) << "Configuring scene for" << connectingSceneSession
+        << "with options" << options;
+
+    UISceneConfiguration *sceneConfig = connectingSceneSession.configuration;
+    sceneConfig.delegateClass = QIOSWindowSceneDelegate.class;
+    return sceneConfig;
+}
+
+@end
+
+@implementation QIOSWindowSceneDelegate
+
+- (void)scene:(UIScene *)scene willConnectToSession:(UISceneSession *)session
+       options:(UISceneConnectionOptions *)connectionOptions
+{
+    Q_UNUSED(session);
+    qCDebug(lcQpaWindowScene) << "Connecting" << scene << "to" << session;
+
+    if (![scene isKindOfClass:[UIWindowScene class]])
+        return;
+
+    UIWindowScene *windowScene = static_cast<UIWindowScene *>(scene);
+    self.window = [[[QUIWindow alloc] initWithWindowScene:windowScene] autorelease];
+    [self configureWindowForScene:windowScene];
+
+    if (connectionOptions.URLContexts.count > 0)
+        [self scene:scene openURLContexts:connectionOptions.URLContexts];
+}
+
+- (void)configureWindowForScene:(UIWindowScene *)windowScene
+{
+    if (!qGuiApp)
+        return;
+
+
+    QIOSScreen *screen = nullptr;
+    for (QScreen *qScreen : qGuiApp->screens()) {
+        QIOSScreen *platformScreen = static_cast<QIOSScreen *>(qScreen->handle());
+        if (platformScreen->uiScreen() == windowScene.screen) {
+            screen = platformScreen;
+            break;
+        }
+    }
+
+    if (!screen) {
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 50 * NSEC_PER_MSEC), dispatch_get_main_queue(), ^{
+            [self configureWindowForScene:windowScene];
+        });
+        return;
+    }
+
+    self.window.rootViewController = [[[QIOSViewController alloc]
+        initWithWindow:self.window andScreen:screen] autorelease];
+    [self.window makeKeyAndVisible];
+}
+
+- (void)scene:(UIScene *)scene openURLContexts:(NSSet<UIOpenURLContext *> *)URLContexts
+{
+    Q_UNUSED(scene);
+    QIOSIntegration *iosIntegration = QIOSIntegration::instance();
+    if (!iosIntegration)
+        return;
+
+    QIOSServices *iosServices = static_cast<QIOSServices *>(iosIntegration->services());
+    for (UIOpenURLContext *urlContext in URLContexts)
+        iosServices->handleUrl(QUrl::fromNSURL(urlContext.URL));
+}
+
+@end
+
+@implementation QIOSApplicationDelegate (URLHandling)
+
 - (BOOL)application:(UIApplication *)application continueUserActivity:(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray<id<UIUserActivityRestoring>> *restorableObjects))restorationHandler
 {
     Q_UNUSED(application);
@@ -87,4 +169,3 @@
 }
 
 @end
-
diff --git a/src/plugins/platforms/ios/qiosglobal.h b/src/plugins/platforms/ios/qiosglobal.h
--- a/src/plugins/platforms/ios/qiosglobal.h
+++ b/src/plugins/platforms/ios/qiosglobal.h
@@ -50,6 +50,7 @@
 Q_DECLARE_LOGGING_CATEGORY(lcQpaApplication);
 Q_DECLARE_LOGGING_CATEGORY(lcQpaInputMethods);
 Q_DECLARE_LOGGING_CATEGORY(lcQpaWindow);
+Q_DECLARE_LOGGING_CATEGORY(lcQpaWindowScene);
 
 #if !defined(QT_NO_DEBUG)
 #define qImDebug \
@@ -69,8 +70,9 @@
 #endif
 
 int infoPlistValue(NSString* key, int defaultValue);
+UIView *rootViewForScreen(const QPlatformScreen *screen);
 
 QT_END_NAMESPACE
 


diff --git a/src/plugins/platforms/ios/qiosglobal.mm b/src/plugins/platforms/ios/qiosglobal.mm
--- a/src/plugins/platforms/ios/qiosglobal.mm
+++ b/src/plugins/platforms/ios/qiosglobal.mm
@@ -49,6 +49,7 @@
 Q_LOGGING_CATEGORY(lcQpaApplication, "qt.qpa.application");
 Q_LOGGING_CATEGORY(lcQpaInputMethods, "qt.qpa.input.methods");
 Q_LOGGING_CATEGORY(lcQpaWindow, "qt.qpa.window");
+Q_LOGGING_CATEGORY(lcQpaWindowScene, "qt.qpa.window.scene");
 
 bool isQtApplication()
 {
@@ -121,6 +122,50 @@
     return value ? [value intValue] : defaultValue;
 }
 
+UIView *rootViewForScreen(const QPlatformScreen *screen)
+{
+    Q_ASSERT(screen);
+
+    const auto *iosScreen = static_cast<const QIOSScreen *>(screen);
+    UIView *fallbackRootView = nil;
+    for (UIScene *scene in [qt_apple_sharedApplication().connectedScenes allObjects]) {
+        if (![scene isKindOfClass:UIWindowScene.class])
+            continue;
+
+        UIWindowScene *windowScene = static_cast<UIWindowScene *>(scene);
+        // iOS 27 can expose a different UIScreen instance to Qt; keep looking
+        // for the scene-owned Qt window and use it as a fallback below.
+
+        UIWindow *uiWindow = nil;
+        id sceneDelegate = windowScene.delegate;
+        if ([sceneDelegate respondsToSelector:@selector(window)]) {
+            UIWindow *delegateWindow = [sceneDelegate window];
+            if ([delegateWindow isKindOfClass:[QUIWindow class]])
+                uiWindow = delegateWindow;
+        }
+        if (!uiWindow && [windowScene.keyWindow isKindOfClass:[QUIWindow class]])
+            uiWindow = windowScene.keyWindow;
+        if (!uiWindow) {
+            for (UIWindow *window in windowScene.windows) {
+                if ([window isKindOfClass:[QUIWindow class]]) {
+                    uiWindow = window;
+                    break;
+                }
+            }
+        }
+        if (!uiWindow)
+            continue;
+        UIView *rootView = uiWindow.rootViewController.view;
+        if (windowScene.screen == iosScreen->uiScreen())
+            return rootView;
+        if (!fallbackRootView)
+            fallbackRootView = rootView;
+
+    }
+
+    return fallbackRootView;
+}
+
 QT_END_NAMESPACE
 
 // -------------------------------------------------------------------------
diff --git a/src/plugins/platforms/ios/qiosscreen.h b/src/plugins/platforms/ios/qiosscreen.h
--- a/src/plugins/platforms/ios/qiosscreen.h
+++ b/src/plugins/platforms/ios/qiosscreen.h
@@ -88,9 +88,8 @@
     void deliverUpdateRequests() const;
 
     UIScreen *m_uiScreen;
-    UIWindow *m_uiWindow;
     QRect m_geometry;
     QRect m_availableGeometry;
     int m_depth;


diff --git a/src/plugins/platforms/ios/qiosscreen.mm b/src/plugins/platforms/ios/qiosscreen.mm
--- a/src/plugins/platforms/ios/qiosscreen.mm
+++ b/src/plugins/platforms/ios/qiosscreen.mm
@@ -251,7 +251,6 @@
 QIOSScreen::QIOSScreen(UIScreen *screen)
     : QPlatformScreen()
     , m_uiScreen(screen)
-    , m_uiWindow(0)
     , m_orientationListener(0)
 {
     QString deviceIdentifier = deviceModelIdentifier();
@@ -284,21 +283,6 @@
         // External display, hard to say
         m_depth = 24;
         m_physicalDpi = 96;
-    }
-
-    if (!qt_apple_isApplicationExtension()) {
-        for (UIWindow *existingWindow in qt_apple_sharedApplication().windows) {
-            if (existingWindow.screen == m_uiScreen) {
-                m_uiWindow = [m_uiWindow retain];
-                break;
-            }
-        }
-
-        if (!m_uiWindow) {
-            // Create a window and associated view-controller that we can use
-            m_uiWindow = [[QUIWindow alloc] initWithFrame:[m_uiScreen bounds]];
-            m_uiWindow.rootViewController = [[[QIOSViewController alloc] initWithQIOSScreen:this] autorelease];
-        }
     }
 
     updateProperties();
@@ -313,7 +297,6 @@
     [m_displayLink invalidate];
 
     [m_orientationListener release];
-    [m_uiWindow release];
 }
 
 QString QIOSScreen::name() const
@@ -333,38 +316,14 @@
 
     m_geometry = QRectF::fromCGRect(m_uiScreen.bounds).toRect();
 
-    // The application frame doesn't take safe area insets into account, and
-    // the safe area insets are not available before the UIWindow is shown,
-    // and do not take split-view constraints into account, so we have to
-    // combine the two to get the correct available geometry.
-    QRect applicationFrame = QRectF::fromCGRect(m_uiScreen.qt_applicationFrame).toRect();
-    UIEdgeInsets safeAreaInsets = m_uiWindow.qt_safeAreaInsets;
-    m_availableGeometry = m_geometry.adjusted(safeAreaInsets.left, safeAreaInsets.top,
-        -safeAreaInsets.right, -safeAreaInsets.bottom).intersected(applicationFrame);
+    // Safe area margins and the maximized state are handled by QIOSWindow.
+    // The scene-owned UIWindow does not exist yet while QGuiApplication starts.
+    m_availableGeometry = m_geometry;
 
 #ifndef Q_OS_TVOS
-    if (m_uiScreen == [UIScreen mainScreen]) {
-        QIOSViewController *qtViewController = [m_uiWindow.rootViewController isKindOfClass:[QIOSViewController class]] ?
-            static_cast<QIOSViewController *>(m_uiWindow.rootViewController) : nil;
-
-        if (qtViewController.lockedOrientation) {
-            Q_ASSERT(!qt_apple_isApplicationExtension());
-
-            // Setting the statusbar orientation (content orientation) on will affect the screen geometry,
-            // which is not what we want. We want to reflect the screen geometry based on the locked orientation,
-            // and adjust the available geometry based on the repositioned status bar for the current status
-            // bar orientation.
-
-            Qt::ScreenOrientation statusBarOrientation = toQtScreenOrientation(
-                UIDeviceOrientation(qt_apple_sharedApplication().statusBarOrientation));
-
-            Qt::ScreenOrientation lockedOrientation = toQtScreenOrientation(UIDeviceOrientation(qtViewController.lockedOrientation));
-            QTransform transform = transformBetween(lockedOrientation, statusBarOrientation, m_geometry).inverted();
-
-            m_geometry = transform.mapRect(m_geometry);
-            m_availableGeometry = transform.mapRect(m_availableGeometry);
-        }
-    }
+    // The old implementation queried a UIWindow owned by QIOSScreen here.
+    // With UIScene, orientation is owned by the scene's view controller and
+    // is updated from QIOSViewController instead.
 #endif
 
     if (m_geometry != previousGeometry) {
@@ -537,26 +496,30 @@
     if (window && ![reinterpret_cast<id>(window) isKindOfClass:[UIView class]])
         return QPixmap();
 
-    UIView *view = window ? reinterpret_cast<UIView *>(window) : m_uiWindow;
+    UIView *view = window ? reinterpret_cast<UIView *>(window) : rootViewForScreen(this);
+    if (!view || !view.window)
+        return QPixmap();
 
+    UIWindow *uiWindow = view.window;
+
     if (width < 0)
         width = qMax(view.bounds.size.width - x, CGFloat(0));
     if (height < 0)
         height = qMax(view.bounds.size.height - y, CGFloat(0));
 
-    CGRect captureRect = [m_uiWindow convertRect:CGRectMake(x, y, width, height) fromView:view];
-    captureRect = CGRectIntersection(captureRect, m_uiWindow.bounds);
+    CGRect captureRect = [uiWindow convertRect:CGRectMake(x, y, width, height) fromView:view];
+    captureRect = CGRectIntersection(captureRect, uiWindow.bounds);
 
     UIGraphicsBeginImageContextWithOptions(captureRect.size, NO, 0.0);
     CGContextRef context = UIGraphicsGetCurrentContext();
     CGContextTranslateCTM(context, -captureRect.origin.x, -captureRect.origin.y);
 
-    // Draws the complete view hierarchy of m_uiWindow into the given rect, which
-    // needs to be the same aspect ratio as the m_uiWindow's size. Since we've
+    // Draws the complete view hierarchy of the scene UIWindow into the given rect, which
+    // needs to be the same aspect ratio as the UIWindow's size. Since we've
     // translated the graphics context, and are potentially drawing into a smaller
     // context than the full window, the resulting image will be a subsection of the
     // full screen.
-    [m_uiWindow drawViewHierarchyInRect:m_uiWindow.bounds afterScreenUpdates:NO];
+    [uiWindow drawViewHierarchyInRect:uiWindow.bounds afterScreenUpdates:NO];
 
     UIImage *screenshot = UIGraphicsGetImageFromCurrentImageContext();
     UIGraphicsEndImageContext();
@@ -571,9 +534,10 @@
 
 UIWindow *QIOSScreen::uiWindow() const
 {
-    return m_uiWindow;
+    UIView *rootView = rootViewForScreen(this);
+    return rootView.window;
 }
 
 #include "moc_qiosscreen.cpp"


diff --git a/src/plugins/platforms/ios/qiosviewcontroller.h b/src/plugins/platforms/ios/qiosviewcontroller.h
--- a/src/plugins/platforms/ios/qiosviewcontroller.h
+++ b/src/plugins/platforms/ios/qiosviewcontroller.h
@@ -48,7 +48,7 @@
 
 @interface QIOSViewController : UIViewController
 
-- (instancetype)initWithQIOSScreen:(QT_PREPEND_NAMESPACE(QIOSScreen) *)screen;
+- (instancetype)initWithWindow:(UIWindow *)window andScreen:(QT_PREPEND_NAMESPACE(QIOSScreen) *)screen;
 - (void)updateProperties;
 
 #ifndef Q_OS_TVOS
diff --git a/src/plugins/platforms/ios/qiosviewcontroller.mm b/src/plugins/platforms/ios/qiosviewcontroller.mm
--- a/src/plugins/platforms/ios/qiosviewcontroller.mm
+++ b/src/plugins/platforms/ios/qiosviewcontroller.mm
@@ -58,6 +58,7 @@
 // -------------------------------------------------------------------------
 
 @interface QIOSViewController ()
+@property (nonatomic, assign) UIWindow *window;
 @property (nonatomic, assign) QPointer<QT_PREPEND_NAMESPACE(QIOSScreen)> platformScreen;
 @property (nonatomic, assign) BOOL changingOrientation;
 @end
@@ -122,17 +123,17 @@
 {
     Q_UNUSED(subview);
 
-    QT_PREPEND_NAMESPACE(QIOSScreen) *screen = self.qtViewController.platformScreen;
+    UIWindow *uiWindow = self.qtViewController.window;
 
-    // The 'window' property of our view is not valid until the window
-    // has been shown, so we have to access it through the QIOSScreen.
-    UIWindow *uiWindow = screen->uiWindow();
-
-    if (uiWindow.hidden) {
-        // Associate UIWindow to screen and show it the first time a QWindow
-        // is mapped to the screen. For external screens this means disabling
-        // mirroring mode and presenting alternate content on the screen.
-        uiWindow.screen = screen->uiScreen();
+    if (uiWindow && uiWindow.hidden) {
+        // Show the scene-owned UIWindow the first time a QWindow is mapped.
         uiWindow.hidden = NO;
+        if ([subview isKindOfClass:[QUIView class]]) {
+            QUIView *view = static_cast<QUIView *>(subview);
+            dispatch_async(dispatch_get_main_queue(), ^{
+                if (view.window == uiWindow)
+                    [view sendUpdatedExposeEvent];
+            });
+        }
     }
 }
@@ -254,9 +248,10 @@
 @synthesize preferredStatusBarStyle;
 #endif
 
-- (instancetype)initWithQIOSScreen:(QT_PREPEND_NAMESPACE(QIOSScreen) *)screen
+- (instancetype)initWithWindow:(UIWindow *)window andScreen:(QT_PREPEND_NAMESPACE(QIOSScreen) *)screen
 {
     if (self = [self init]) {
+        self.window = window;
         self.platformScreen = screen;
 
         self.changingOrientation = NO;
@@ -318,6 +313,26 @@
             name:UIApplicationDidChangeStatusBarOrientationNotification
             object:qt_apple_sharedApplication()];
 #endif
+
+    // The QML top-level window can be created before UIKit connects the scene.
+    // Reparent those already-created Qt windows into this scene's desktop view.
+    for (auto *window : qGuiApp->topLevelWindows()) {
+        if (window->screen()->handle() != self.platformScreen)
+            continue;
+        if (auto *platformWindow = window->handle())
+            platformWindow->setParent(nullptr);
+    }
+
+    // The scene may have been connected after the Qt window was created.
+    // Re-apply the window state now that the view has a real UIWindow parent,
+    // so fullscreen geometry and exposure are recalculated from scene bounds.
+    for (auto *window : qGuiApp->topLevelWindows()) {
+        if (window->screen()->handle() != self.platformScreen)
+            continue;
+        if (auto *platformWindow = window->handle())
+            platformWindow->setWindowState(window->windowStates());
+    }
+    [self.view setNeedsLayout];
 }
 
 - (void)viewDidUnload
diff --git a/src/plugins/platforms/ios/qioswindow.mm b/src/plugins/platforms/ios/qioswindow.mm
--- a/src/plugins/platforms/ios/qioswindow.mm
+++ b/src/plugins/platforms/ios/qioswindow.mm
@@ -58,3 +58,4 @@
 #include <QtDebug>
+#include <QtCore/QPointer>

 QT_BEGIN_NAMESPACE
@@ -242,5 +242,8 @@
 bool QIOSWindow::isExposed() const
 {
-    return qApp->applicationState() != Qt::ApplicationSuspended
-        && window()->isVisible() && !window()->geometry().isEmpty();
+    return qApp->applicationState() != Qt::ApplicationSuspended
+        && window()->isVisible() && !window()->geometry().isEmpty()
+        && m_view.window && m_view.window.rootViewController
+        && m_view.window.rootViewController.view.window == m_view.window
+        && [m_view isDescendantOfView:m_view.window.rootViewController.view];
 }
@@ -265,15 +265,68 @@
         // in applyGeometry(), as we don't want to artificially limit window
         // placement "outside" of the screen bounds if that's what the user wants.
 
-        QRect uiWindowBounds = QRectF::fromCGRect(m_view.window.bounds).toRect();
-        QRect fullscreenGeometry = screen()->geometry().intersected(uiWindowBounds);
-        QRect maximizedGeometry = window()->flags() & Qt::MaximizeUsingFullscreenGeometryHint ?
-            fullscreenGeometry : screen()->availableGeometry().intersected(uiWindowBounds);
+        QRect uiWindowBounds = m_view.window ? QRectF::fromCGRect(m_view.window.bounds).toRect() : QRect();
+        if (NSProcessInfo.processInfo.iOSAppOnMac) {
+            // iPad apps running on macOS use a scaled UIScreen coordinate system.
+            // The scene UIWindow is the authoritative coordinate space.
+            applyGeometry(uiWindowBounds);
+        } else {
+            // On iPhone Duo, Qt's UIScreen can describe one panel while the
+            // scene-owned UIWindow spans the open device. Use the scene window
+            // as the authoritative geometry whenever it is available.
+            QRect fullscreenGeometry = uiWindowBounds.isEmpty()
+                ? screen()->geometry() : uiWindowBounds;
+            QRect maximizedGeometry = uiWindowBounds.isEmpty()
+                ? screen()->availableGeometry() : uiWindowBounds;
+            if (m_view.window) {
+                UIScreen *uiScreen = m_view.window.windowScene.screen;
+                UIEdgeInsets safeAreaInsets = m_view.window.safeAreaInsets;
+                const CGSize screenBounds = uiScreen ? uiScreen.bounds.size : CGSizeZero;
+                const CGFloat screenLongSide = MAX(screenBounds.width, screenBounds.height);
+                const CGFloat screenShortSide = MIN(screenBounds.width, screenBounds.height);
+                const BOOL isIPhoneDuo =
+                    m_view.window.windowScene.traitCollection.userInterfaceIdiom == UIUserInterfaceIdiomPhone
+                    && ((screenLongSide >= 1400 && screenLongSide <= 1450
+                         && screenShortSide >= 980 && screenShortSide <= 1020)
+                        || (screenLongSide >= 650 && screenLongSide <= 700
+                            && screenShortSide >= 450 && screenShortSide <= 500)
+                        || (screenLongSide >= 930 && screenLongSide <= 970
+                            && screenShortSide >= 650 && screenShortSide <= 690));
+                const BOOL isIPhoneDuoInnerDisplay =
+                    isIPhoneDuo
+                    && m_view.window.windowScene.traitCollection.horizontalSizeClass == UIUserInterfaceSizeClassRegular;
+                if (uiScreen) {
+                    if (!isIPhoneDuoInnerDisplay
+                        && m_view.window.bounds.size.width == uiScreen.bounds.size.width)
+                        maximizedGeometry.adjust(safeAreaInsets.left, 0,
+                            -safeAreaInsets.right, 0);
+                    if (m_view.window.bounds.size.height == uiScreen.bounds.size.height)
+                        maximizedGeometry.adjust(0, safeAreaInsets.top,
+                            0, -safeAreaInsets.bottom);
+                }
+
+                // Keep the Duo scene width (the right system rail stays
+                // outside QZ) but let full-screen content reach the
+                // bottom of the physical display.
+                if (isIPhoneDuo && !uiWindowBounds.isEmpty()) {
+                    fullscreenGeometry.setX(uiWindowBounds.x());
+                    fullscreenGeometry.setY(uiWindowBounds.y());
+                    fullscreenGeometry.setWidth(uiWindowBounds.width());
+                    fullscreenGeometry.setHeight(uiWindowBounds.height() + safeAreaInsets.bottom);
+                    maximizedGeometry = fullscreenGeometry;
+                }
+
+                if (!isIPhoneDuo) {
+                    fullscreenGeometry = fullscreenGeometry.intersected(uiWindowBounds);
+                    maximizedGeometry = maximizedGeometry.intersected(uiWindowBounds);
+                }
+            }
 
-        if (state & Qt::WindowFullScreen)
-            applyGeometry(fullscreenGeometry);
-        else
-            applyGeometry(maximizedGeometry);
+            if (state & Qt::WindowFullScreen)
+                applyGeometry(fullscreenGeometry);
+            else
+                applyGeometry(maximizedGeometry);
+        }
     } else {
         applyGeometry(m_normalGeometry);
     }
@@ -282,8 +291,16 @@
 void QIOSWindow::setParent(const QPlatformWindow *parentWindow)
 {
     UIView *parentView = parentWindow ? reinterpret_cast<UIView *>(parentWindow->winId())
-        : isQtApplication() ? static_cast<QIOSScreen *>(screen())->uiWindow().rootViewController.view : 0;
+        : rootViewForScreen(screen());

-    [parentView addSubview:m_view];
+    if (parentView)
+        [parentView addSubview:m_view];
+    else if (!parentWindow) {
+        QPointer<QIOSWindow> window = this;
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 50 * NSEC_PER_MSEC), dispatch_get_main_queue(), ^{
+            if (window && !window->m_view.superview)
+                window->setParent(nullptr);
+        });
+    }
 }
 
@@ -290,10 +290,15 @@
 void QIOSWindow::requestActivateWindow()
 {
     // Note that several windows can be active at the same time if they exist in the same
     // hierarchy (transient children). But only one window can be QGuiApplication::focusWindow().
     // Dispite the name, 'requestActivateWindow' means raise and transfer focus to the window:
     if (blockedByModal())
         return;
 
+    // The QWindow can become visible before UIKit has connected the scene.
+    // Activation is retried when the view is attached to the scene-owned window.
+    if (!m_view.window)
+        return;
+
     Q_ASSERT(m_view.window);

diff --git a/src/plugins/platforms/ios/qioscontext.mm b/src/plugins/platforms/ios/qioscontext.mm
--- a/src/plugins/platforms/ios/qioscontext.mm
+++ b/src/plugins/platforms/ios/qioscontext.mm
@@ -138,7 +138,15 @@
 bool QIOSContext::makeCurrent(QPlatformSurface *surface)
 {
     Q_ASSERT_IS_GL_SURFACE(surface);
 
+    // QSG may start its render thread before UIScene has attached the Qt view
+    // to the scene-owned UIWindow. Do not touch CAEAGLLayer until UIKit has
+    // completed that attachment; the expose event from QUIView::didMoveToWindow
+    // will make the scene renderable again.
+    if (surface->surface()->surfaceClass() == QSurface::Window
+        && !static_cast<QIOSWindow *>(surface)->isExposed())
+        return false;
+
     if (!verifyGraphicsHardwareAvailability())
         return false;

diff --git a/src/plugins/platforms/ios/quiview.mm b/src/plugins/platforms/ios/quiview.mm
--- a/src/plugins/platforms/ios/quiview.mm
+++ b/src/plugins/platforms/ios/quiview.mm
@@ -230,3 +230,17 @@
 }
 
+- (void)didMoveToWindow
+{
+    [super didMoveToWindow];
+
+    // A QWindow can be created before UIScene attaches its view hierarchy.
+    // Re-emit exposure after attachment so QSG retries rendering now that
+    // CAEAGLLayer has a real scene-owned UIWindow.
+    if (self.window && self.platformWindow) {
+        [self setNeedsLayout];
+        [self setNeedsDisplay];
+        [self sendUpdatedExposeEvent];
+    }
+}
+
 - (void)displayLayer:(CALayer *)layer
